## Viterbi Algorithm

### Intialisation

In [1]:
import numpy as np

states = ["S1 (/h/)", "S2 (/e/)", "S3 (/l/)", "S4 (/o/)"]
observations = ["O1", "O2", "O3", "O4"]

A = np.array([
    [0.0, 0.7, 0.3, 0.0],
    [0.0, 0.2, 0.6, 0.2],
    [0.0, 0.0, 0.3, 0.7],
    [0.0, 0.0, 0.1, 0.9]
])

B = np.array([
    [0.6, 0.2, 0.1, 0.1],  
    [0.1, 0.7, 0.1, 0.1],  
    [0.1, 0.1, 0.6, 0.2],  
    [0.2, 0.1, 0.2, 0.5]   
])

pi = np.array([1.0, 0.0, 0.0, 0.0])

O_seq = [0, 1, 2, 3]  

In [2]:
def viterbi(O_seq, A, B, pi):
    n_states = A.shape[0]
    T = len(O_seq)
    
    dp = np.zeros((n_states, T))  
    path = np.zeros((n_states, T), dtype=int)
    
    dp[:, 0] = pi * B[:, O_seq[0]]
    
    for t in range(1, T):
        for j in range(n_states):
            probabilities = dp[:, t-1] * A[:, j] * B[j, O_seq[t]]
            dp[j, t] = np.max(probabilities)
            path[j, t] = np.argmax(probabilities)
    
    final_state = np.argmax(dp[:, -1])
    max_prob = dp[final_state, -1]
    
    best_path = []
    current_state = final_state
    for t in range(T-1, -1, -1):
        best_path.insert(0, current_state)
        current_state = path[current_state, t]
    
    return best_path, max_prob

In [3]:
best_path_indices, max_prob = viterbi(O_seq, A, B, pi)
best_path = [states[i] for i in best_path_indices]
print("Most Likely Sequence of States:", best_path)
print("Probability of the Most Likely Sequence:", max_prob)

Most Likely Sequence of States: ['S1 (/h/)', 'S2 (/e/)', 'S3 (/l/)', 'S4 (/o/)']
Probability of the Most Likely Sequence: 0.03704399999999999


### Inference

The Viterbi Algorithm identifies the sequence /h/, /e/, /l/, /o/ as the most probable phoneme sequence for the given observation sequence [O1, O2, O3, O4].

The computed probability of the sequence is approximately 0.037, which is the highest likelihood among all possible phoneme sequences.

The alignment of the states and observations is consistent with the expected phoneme transitions, showcasing the capability of the Viterbi Algorithm to decode hidden states in a sequential manner.

## END